### PR TITLE
Stop collecting disposables if changed token encountered

### PR DESCRIPTION
**Pull request by mapogolions**

### CODE REVIEW

**Comment by mapogolions** [comment]
@eerhardt 
####  Traditional flow:
We usually pass a list of tokens that have not yet been changed to the constructor of the composite token.
```
var compositeToken = new CompositeChangeToken(token1, token2, ..., tokenN);
```
where: `token1.HasChanged is False, token2.HasChanged is False, ..., tokenN.HasChanged is False`

Then we register some callback for the composite token. 
```
compositeToken.Register(...);
```
at this moment `compositeToken.HasChanged is False`

At the output, we have a construct in which the created composite token has not yet been changed 
and each of the passed tokens (token1, token2, ..., tokenN) can initiate a change in the composite token in the future. 

If cancellation is requested for one of these tokens 
```
tokenSource1.Cancel()
```
the static OnChange method will be called. [What it does](https://github.com/dotnet/runtime/blob/88c2a664c211d6562a1bc89ab15a890608181bdf/src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs#L121):
1) lock and cancel composite token
2) Dispose all disposables (registrations)

If cancellation is then requested for the second passed token 
```
tokenSource2.Cancel()
```
The OnChange method will no longer be called as all disposables(registrations) have been disposed.


#### Issue flow:
We also pass a list of tokens, but one of them has already been changed.
```
var compositeToken = new CompositeChnageToken(token1, token2, ..., tokenN);
```
where:  __`token1.HasChanged is True`__, `token2.HasChanged is False, ..., tokenN.HasChanged is False`

Then we register some callback for the composite token
```
compositeToken.Register(...)
```
The static method OnChange will be called at the moment of registration by the line above.
As result, we have a changed composite token (e.g. compositeToken.HasChanged is True), but we also have a list of disposables
that hasn't been disposed.
And If cancellation is then requested for the second passed token
```
tokenSource2.Cancel()
```
It will trigger the OnChange method again. 

This PIR is trying to remove redundant work - if we know that the composite token has changed, why do we still keep registering a callback for each token in order to be able to change the composite token in the future

File: src/libraries/Microsoft.Extensions.Primitives/tests/CompositeChangeTokenTest.cs [hunk 1]
```diff
@@ -148,5 +148,66 @@ public async Task RegisteredCallbackGetsInvokedExactlyOnce_WhenMultipleConcurren
...
+    }
+
+    internal class ProxyCancellationChangeToken : IChangeToken
+    {
+        private readonly CancellationChangeToken _cancellationChangeToken;
+        private readonly Action _disposing;
+
+        public ProxyCancellationChangeToken(CancellationToken cancellationToken, Action disposing)
+        {
+            _cancellationChangeToken = new(cancellationToken);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/tests/CompositeChangeTokenTest.cs:182]:
```suggestion
            _cancellationChangeToken = new CancellationChangeToken(cancellationToken);
```

Just like `var` without the right-hand side being obvious on the Type, we don't use `new()` without the left-hand side being obvious on the Type.

File: src/libraries/Microsoft.Extensions.Primitives/tests/CompositeChangeTokenTest.cs [hunk 2]
```diff
@@ -148,5 +148,66 @@ public async Task RegisteredCallbackGetsInvokedExactlyOnce_WhenMultipleConcurren
...
+            _disposing = disposing;
+        }
+        public bool ActiveChangeCallbacks => _cancellationChangeToken.ActiveChangeCallbacks;
+
+        public bool HasChanged => _cancellationChangeToken.HasChanged;
+
+        public IDisposable RegisterChangeCallback(Action<object?> callback, object? state)
+        {
+            var registration = _cancellationChangeToken.RegisterChangeCallback(callback, state);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/tests/CompositeChangeTokenTest.cs:191]:
```suggestion
            IDisposable registration = _cancellationChangeToken.RegisterChangeCallback(callback, state);
```

**Review by eerhardt — APPROVED** [review]
LGTM. Just a couple minor code style changes.

FYI @davidfowl - in case you have any concerns here.

**Comment by mapogolions** [comment]
@eerhardt Thanks for review. I am still working on my English. I would appreciate any of your suggestions for renaming local variables, classes, functions.

**Comment by eerhardt** [comment]
> I would appreciate any of your suggestions for renaming local variables, classes, functions.

I thought they all looked fine.

**Comment by eerhardt** [comment]
Test failure was #65791.

**Pull request merged**